### DISCUSSION TITLE
Modularity?

### DISCUSSION THREAD

**Discussion by dr-phill** [blackears/svgSalamander#95] [Ideas]
Hi there,
the  descriptions make this sound like a great match for my needs, but I do not see a module info.java file here....   I am forced by circumstance only to depend upon modular jars so the lack is a blocker for me.

Are you planning on 'going modular'?  If not I might see what I might achieve, but I am likely to get the exposure wrong which could have unintended consequences.

- Phill

**Comment by blackears** [comment]
I'm not very familiar with modularity in Java (this library predates it by quite a bit).  

I've updated the source to use Java 9 and included a module-info.java file.  Please check this out and let me know if it works and the module exports are set up correctly.

**Reply by dr-phill** [reply]
Wow, that was fast. Thanks

Your module-info looks good, but I cannot test it.  I do not have the correct build environment set up. I have not seen your build environment before - have you any links to get me started?   Maybe I will have to resort to javac?

**Comment by blackears** [comment]
It uses Maven to build, so if you have Maven installed you can build it that way.  I also use the Netbeans IDE, so it should be buildable if you import it into that.  Although most modern IDEs should be able to handle Maven projects.

**Reply by dr-phill** [reply]
OK, cool.  I have never used maven as a build system.  I use gradle.  No problems, probably about time I learned.
I guess I execute a maven command from the directory with a pom?......
.....
'mvn package' gets me a jar file (svgSalamander-1.1.4.jar).  I guess I am good to try to incorporate that into my build next.
 
Thanks, I will let you know how I get on.

**Comment by dr-phill** [comment]
Hmmmm
`mvn clean install`
places the jar in my local maven repository, but the jar looks bogus.
If I unpack it with the (macOS) packagemanager I get a file called
`svgSalamander-1.1.4.jar.cpgz`
unpacking that creates another
`svgSalamander-1.1.4.jar`
This is the sign of a 'corrupt zip file'. - but it may not be corrupt just not recognised by macOS.
What platform do you normally build on?  I can easily try linux (ubuntu) but might struggle to dig out windows from somewhere....

**Comment by blackears** [comment]
You might want to try ```mvn package``` instead.  I'm not very familiar with Maven and lean on my IDE to do much of the work.

My development machine runs Windows 10.

**Comment by dr-phill** [comment]
OK, got it working now... the 'corruption' of the published jar was a red herring. I can publish to mavenLocal and use within my project.
Odd results though - probably something that I am doing wrong.  I will investigate further.
But your module-info seems to work fine.

**Comment by dr-phill** [comment]
OK some odd results.  At first I thought that it was my home-brewed sag file, but it seems to break with one of your test files.
<img width="505" alt="Screenshot 2022-02-28 at 10 12 31" src="https://user-images.githubusercontent.com/95749004/155965210-d35ff934-903c-4614-b507-2ca28975e8fb.png">

my code:
```
override fun paint(g: Graphics?) {
    super.paint(g)
    val gg = g as? Graphics2D ?: return
    val svgu = SVGUniverse()
    val file = File("./jira51-textLength.svg")
    val uri = file.toPath().normalize().toUri()
    svgu.loadSVG(uri.toURL())
    val svgD = svgu.getDiagram(uri)
    svgD.render(gg)
}
```
Apologies for the Kotlin code ... but it should be easy to guess the java...
The only thing that I can think of here is that somehow the resolution of my iMac is throwing the calculations off.

**Reply by blackears** [reply]
I don't think the lengthAdjust tag was ever implemented.  It's read in, but it looks like it is not used when calculating the text layout.

**Comment by dr-phill** [comment]
OK, I am not really worried about that render, but this one is important to me.  I did not want to post this one initially as it is one of my auto-generated images though it does render correctly in Firefox, safari, chrome etc....
Here is the rendering I get in a java app using salamander:
<img width="478" alt="image" src="https://user-images.githubusercontent.com/95749004/155983961-2d1a2577-00c6-418e-9cb4-4cf5d4a695f6.png">
Here is the rendering I get in chrome:
<img width="457" alt="image" src="https://user-images.githubusercontent.com/95749004/155984414-bec54c60-29bf-4cd5-9ccf-47e7c5eb709b.png">

Here is my sag:
```
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<svg width="455" height="312" xmlns="http://www.w3.org/2000/svg">
<rect fill="#f0f0f0" stroke="#000" width="455" height="312"/>
<g transform="translate(229,248)">
    <rect width="196" height="34"  
          rx="17" ry="17"  fill="#ffffff"  
          stroke-width="1" stroke="#000000"/>
</g>
<g transform="translate(230,249)">
    <rect width="194" height="32"  fill-opacity="0"  stroke-width="0" stroke="#ffffff"/>
</g>
<g transform="translate(235,254)">
    <svg width="184" height="22">
         <text x="50%" y="50%" font-size="18" 
               font-style="normal" font-weight="normal"  fill="#000000" 
               dominant-baseline="central" text-anchor="middle">
            Rounded Border
        </text> 
    </svg>
</g>
</svg>

```

I am quite happy to accept that my auto-generated code is bogus, but I need to be able to do two things 
(1) svg --> java image (which salamander does)
(2) java --> svg (which I am not sure salamander does)
If salamander does (2) then I am not worried about my home-spun svg generation.

**Reply by blackears** [reply]
I'm not sure what you mean by (2).  SVG Salamander is itself a svg rendering engine that runs in Java.  Do you mean you want some sort of text input field?

I think the problem with your svg is the "50%" is not being processed correctly by the Tspan tag.  That's something else that should be fixed.

**Comment by dr-phill** [comment]
Cool, not my renderer at fault.

What I mean by 2 is that given a diagram composed go java objects (well, kotlin but same difference) I want to convert that to svg.
Currently my diagram elements render in two different ways.

1. to a Graphics2D object (where I hope to use salamander to render a svg to the screen as part of my diagram) 
2.  to my own SvgWriter where I output text in xml format.
```
An example of the second would be:
class RectangleBorder {
override fun write(svgWriter: SvgWriter) {
        svgWriter.fill = backgroundColour
        svgWriter.stroke(borderColour)
        svgWriter.strokeWidth = stroke.toInt()
        svgWriter.rect(x, y, width, height)
}
}
```

```
class SvgWriter {
public fun rect(x: Int, y: Int, width: Int, height: Int) {
      writer.write(
          """
          <g transform="translate($x,$y)">
              <rect width="$width" height="$height" ${fillString()} stroke-width="$strokeWidth" stroke="$stroke"/>
          </g>
          """.trimIndent()
      )
      writer.write("\n")
}
```
Which is obviously a poor solution on many levels.  I am hoping I can use the code in salamander to manipulate higher level concepts that strings.

I looked at Batik (though it is not modular and is way more complicated than salamander) but it did subclass Graphics2D to output direct to an SVG format.  That would be way cool, because I could reuse the paint() method in my code.  But I can live without that.

**Comment by blackears** [comment]
I think I started writing a SVG Salamander exporter a long time ago, but it never was completed.  You could create your own Batik style exporter by overriding Graphics2D - I did that once and discovered overriding it was surprisingly easy.  The trouble with that approach is that you will only be receiving the final write-to-screen events which may or may not be what you want.  Instead you could try walking the tree of nodes in the SVGDiagram and writing them out which should give you a more accurate representation.

When I first designed SVGSalamander I made the mistake of assuming I could represent each of the tags with a plain Java object.  Then I realized that in SVG so many of the different fields could take on multiple of types.  If I were to redo it, I'd make the core structure more closely resemble an XML document which would solve a lot of the problems that it has.

**Reply by dr-phill** [reply]
Thanks for that info.  
I did once override wrap Graphics2D (so I could reuse my display code in swing or on android).  The main problem with that is handling fonts.  
My current solution to outputting svg is a crude version of the tree walker.  But it will have to do for now.
 At least I can ender svg to graphics2d using salamander so that is a problem solved.

Thanks for your help and for your code.  I appreciate it.

Phill

**Comment by dr-phill** [comment]
Hi there,

I have been looking at your code and thinking about possibilities.  How feasible would it be to created a (shallow) tree of SVGElement programatically, then unparse these to an xml document?
The simple elements (rectangles etc) should not pose too much of a problem, but what about the 'relational' elements and things like embedded fonts?

My existing code already does some of this, but without using the intermediate representation. Instead of writing the xml for a rectangle could I create a Rect SVGElement and insert it into parent?  Would the parent then have enough information for me to walk the tree and render a valid svg/xml file?  Maybe you already have this code and I have not spotted it?

Thanks

Phill

**Comment by blackears** [comment]
I've not actually used this code myself for several years, so my memory is a bit fuzzy.  IIRC, the document node structure should follow that of an svg document closely.  All references are stored in a table stored in one of the root elements (SVGUniverse, I think?) so you can look there for your definitions of symbols and fonts.  You probably don't even need to create a copy of the tree first.  Just create a class that recursively walks each node and creates XML for it.

You can manually build the tree by creating elements and adding them.  Make sure to call the build() method so that the internal state of the node is updated to reflect the XML attribute values.

**Comment by dr-phill** [comment]
HI there, thanks for that - I can see how that could be done.  

I have decided, however, to have a crack at implementing an svg renderer in Kotlin.  I have done the easy bits (architecture, basic shapes, colours, solid fills, opacities).  I am using the the spec and your code as references and aiming for a minimum-spec (static, standalone) implementation as that is all I need.

My next task is to understand the text rendering part..... that looks like a bunch of fun.

Thanks for your help, and for your code - it is aiding my understanding immensely.

Phill
